scripts/api.js

```diff
@@ -1,17 +1,30 @@
-/*global $ codeUrl */
-/* exported makeApiAjaxSettings */
+/*global codeUrl */
+/* exported ajax */
 
-function makeApiAjaxSettings(apiUrl, queryParams) {
-    let queryParamString = "";
-    if(queryParams) {
-        queryParamString = "&" + $.param(queryParams);
-    }
-    let settings = {
-        "url": codeUrl + "/api/index.php?url=" + apiUrl + queryParamString,
-        "dataType": "json",
-        "headers": {
-            "X-API-KEY": "SESSION"
-        }
+function ajax(method, apiUrl, queryParams = {}, data = {}) {
+    let url = new URL(`${codeUrl}/api/index.php`);
+    queryParams.url = apiUrl;
+    url.search = new URLSearchParams(queryParams);
+    let options = {
+        headers: {"X-API-KEY": "SESSION"},
+        method: method,
     };
-    return settings;
+    if(method !== "GET") {
+        // POST or PUT
+        options.headers['Content-Type'] = 'application/json';
+        options.body = JSON.stringify(data);
+    }
+    return new Promise(function(resolve, reject) {
+        fetch(url, options)
+            .then(function(response) {
```

Comment:
not that it can happen often, but the fetch promise could need a catch if network interrupted or something so may want to catch the fetch promise too:
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

Reply:
I put in a catch and tried issuing a request with the network disconnected but nothing ever happens. We could perhaps put in a timer to detect network disconnection.
The article suggests that a network error could result from misconfigured server. Can we simulate this to test it?